core/arch/arm/kernel/generic_boot.c

```diff
@@ -821,6 +825,30 @@ static void reset_dt_references(void)
 {
 }
 
+static void carveout_memory_range(void)
+{
+	const struct core_mmu_phys_mem *mem;
+	size_t nelems = 0;
+
+	/* check if nsec ddr had been registered by register_nsec_ddr()*/
+	if (&__start_phys_nsec_ddr_section != &__end_phys_nsec_ddr_section)
+		return;
```

Comment:
`IMSG("Registered ddr overridden by registered nsec ddr");`

Reply:
I find the text a bit cryptic for an info-level (i.e., non-debug) message...

Reply:
ok :stuck_out_tongue:
Maybe assert if both nsec_ddr and ddr_overall are registered? I think a platform should not mess defining inconsistent data.

Reply:
So, stay as it?😛

Reply:
Proposal:
```
/* Platform cannot define nsec_ddr && overall_ddr */
assert(__start_phys_nsec_ddr_section == &__end_phys_nsec_ddr_section);
```

Reply:
The carve out operation will be called when CFG_DT is not defined. If changing it as your said, it mean that optee will not support `register_nsec_ddr()` anymore when CFG_DT is not defined. 
I think what your mean is that platform should not call `register_nsec_ddr()` and `register_ddr()` at same time. We could check that if nsec_ddr and overall_ddr are defined at same time in `init_mem_map()`. What do you think?

Reply:
ok I understand. Maybe this way (below) or in `init_mem_map()` as you suggested.

```c
static void carveout_memory_range(void)
{
	struct core_mmu_phys_mem *mem;
	size_t nelems = (&__end_phys_ddr_overall_section -
			&__start_phys_ddr_overall_section) / sizeof(*mem);

	if (!nelems)
		return;

	/* Platform cannot define nsec_ddr && overall_ddr */
	assert(&__start_phys_nsec_ddr_section == &__end_phys_nsec_ddr_section);

	mem = calloc(nelems, sizeof(*mem));
	if (!mem)
		panic();

	memcpy(mem, &__start_phys_ddr_overall_section, sizeof(*mem) * nelems);
	core_mmu_set_discovered_nsec_ddr((struct core_mmu_phys_mem *)mem,
					nelems);
}
```

Reply:
I prefer to use your method: )